src/compiler_internals.md

```diff
@@ -0,0 +1,37 @@
+
+**AGA** looks very similar to assembly, but with annotation on events, closures, and the dependency graph that the Alan Runtime uses reorganize for concurrency and parallelism benefits.
+
+**AGC** is the binary version of *AGA* and is a near direct translation of the text to a 64-bit little-endian integer packed array, with keywords being given numeric values that are clearly interpretable as 8 UTF-8/ASCII characters.
+
+The compiler contains as a set of AST generators for the various input formats ([ln](https://github.com/alantech/alan/tree/main/compiler/src/ln), [amm](https://github.com/alantech/alan/blob/main/compiler/src/amm.ts), [aga](https://github.com/alantech/alan/blob/main/compiler/src/aga.ts)) with ln using [ANTLR](https://www.antlr.org/) to define its grammar and the other two using our homegrown [lp](https://github.com/alantech/alan/blob/main/compiler/src/lp.ts).
+
+Next, the compiler has the translation stages that go from one input format to another output format ([lntoamm](https://github.com/alantech/alan/tree/main/compiler/src/lntoamm), [ammtojs](https://github.com/alantech/alan/blob/main/compiler/src/ammtojs.ts), [ammtoaga](https://github.com/alantech/alan/blob/main/compiler/src/ammtoaga.ts), [agatoagc](https://github.com/alantech/alan/blob/main/compiler/src/agatoagc.ts)).
+
+While overkill at the moment, the compiler [solves the shortest path from the specified input file to output file and constructs the pipeline to do so on the fly](https://github.com/alantech/alan/blob/main/compiler/src/pipeline.ts). (This is much faster than it sounds and has allowed us to add new inputs/outputs and translation formats quickly.)
+
+The first stage of the compiler dominates with more than 90% of the total compile time, and within that, the ANTLR-generated parser takes over 99% of the time within the `lntoamm` layer. (Speeding this up is very important to us.) Once the AST is generated, the first layer transforms this into a scope-oriented collection of elements and then generates AMM code on a per-event-handler basis, finally it looks for constants in these handlers and hoists them to the global scope and deduplicates them, then emits this in the AMM format.
+
+The next stage of the compiler take the AMM and converts it either into Javascript or AGA. Javascript is a very simple translation, as AMM was designed to look like a "least common denominator scope-based language" which is a very large number of popular languages. The AGA conversion takes a bit more time as the variables need to be assigned memory addresses during their run and nested scopes in event handlers pulled out into "closure handlers", but both of these conversions, both the parsing of AMM and the generation of JS or AGA, take less than 1% of the time the first stage compiler takes.
```

Comment:
- It is worth pointing out that the translation from AMM to JS is simple because they are both event based and this wouldn't be as true for other popular languages
- nit: takes the AMM

Reply:
> It is worth pointing out that the translation from AMM to JS is simple because they are both event based and this wouldn't be as true for other popular languages

Not very much, though? You either go with a purely blocking implementation and the event emit stuff is done with a super-simple shim (push the event and payload to an array and loop over that array for new work once the current work is done) or you use an event library in the language like Python and Java and others have. Even in Javascript I decided to use Node's `EventEmitter` library to make it easier and not going pure Javascript.

I don't think these sorts of details are super important here, since we're unlikely to write another target for a single-threaded garbage-collected language?

> nit: takes the AMM

Will fix.

Reply:
Nevermind on that, going to address the first point, too. I found a way to do it without taking too much extra space.